src/meta/src/hummock/compaction/picker/min_overlap_compaction_picker.rs

```diff
@@ -222,129 +221,114 @@ impl NonOverlapSubLevelPicker {
-            overlap_info.update(sst);
-            select_sst_id_set.insert(sst.sst_id);
-        }
 
+        let mut pick_levels_range = Vec::default();
+        let mut max_select_level_count = 0;
+
+        // Pay attention to the order here: Make sure to select the lowest sub_level to meet the requirements of base compaction. If you break the assumption of this order, you need to redesign it.
+        // TODO: Use binary selection to replace the step algorithm to optimize algorithm complexity
         for (target_index, target_level) in levels.iter().enumerate().skip(1) {
             if target_level.level_type() != LevelType::Nonoverlapping {
                 break;
             }
 
-            // more than 1 sub_level
-            if ret.total_file_count > 1 && ret.total_file_size >= self.max_compaction_bytes
-                || ret.total_file_count >= self.max_file_count as usize
+            if ret
+                .sstable_infos
+                .iter()
+                .filter(|ssts| !ssts.is_empty())
+                .count()
+                > MAX_COMPACT_LEVEL_COUNT
             {
                 break;
             }
 
+            // reset the `basic_overlap_info` with basic sst
+            let mut basic_overlap_info = self.overlap_strategy.create_overlap_info();
+            basic_overlap_info.update(sst);
+
             let mut overlap_files_range =
-                overlap_info.check_multiple_include(&target_level.table_infos);
+                basic_overlap_info.check_multiple_include(&target_level.table_infos);
             if overlap_files_range.is_empty() {
                 overlap_files_range =
-                    overlap_info.check_multiple_overlap(&target_level.table_infos);
+                    basic_overlap_info.check_multiple_overlap(&target_level.table_infos);
             }
 
-            // We allow a layer in the middle without overlap, so we need to continue to
-            // the next layer to search for overlap
-            let mut pending_compact = false;
-            let mut current_level_size = 0;
-            for index in overlap_files_range.start..overlap_files_range.end {
-                let other = &target_level.table_infos[index];
-                if level_handler.is_pending_compact(&other.sst_id) {
-                    pending_compact = true;
-                    break;
-                }
-                overlap_info.update(other);
-                select_sst_id_set.insert(other.sst_id);
-                current_level_size += other.file_size;
+            if overlap_files_range.is_empty() {
+                continue;
             }
 
-            if pending_compact {
-                break;
-            }
+            let mut overlap_levels = vec![];
 
-            let mut extra_overlap_levels = vec![];
+            let mut add_files_size: u64 = 0;
+            let mut add_files_count: usize = 0;
+            let mut finish = false;
+            let mut pending_compact = false;
```

Comment:
nits: It turns out that these two flags are only used to break nested loops. How about using loop labels to make it clearer? [example](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=73e3f036337848a0c750338b878a586d)